File: 2-Regression/5-Multiple Linear Regression/multiple_linear_regression_grade_predict.py
# Multiple Linear Regression

# STEP 1

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
gradesdataset = pd.read_csv('50_Grades_Data.csv')
X = gradesdataset.iloc[:, :-1].values # Removes last col, keeps only independent vars
# Since X will be of type object, input X into console to vew the matrix values
y = gradesdataset.iloc[:, 3].values # Selects last col

# Encoding categorical data
# Encoding the Independent Variable State
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 2] = labelencoder_X.fit_transform(X[:, 2]) # Select state col
onehotencoder = OneHotEncoder(categorical_features = [2])
X = onehotencoder.fit_transform(X).toarray()
# Dependent variables don't need to be encoded


# Avoiding the Dummy Var Trap
# Library will take care of trap but added here as reminder
X = X[:, 1:]

# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Manual feature scaling not needed

# STEP 2

# Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# STEP 3

# Predicting the Test set results
y_pred = regressor.predict(X_test)

######################################################

# Backward Elimination
# STEP 1

# The goal is to find a team of independent variables that have the most statistical significance. 



# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm

# Create column of 1s to associate with constant B zero, so the coloumn of 1s can be thought of as X zero equal to 1
X =  np.append(arr = np.ones((50,1)).astype(int), values = X, axis =1)

# Create matrix only with independent vars that have hight impact
# Start with matrix of features of all ind vars
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
# Create new regressor object
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor with highest p value
regressor_OLS.summary()

# See in console that X2 has p value highest
# Remove X2
X_opt = X[:, [0, 1, 3, 4, 5]]
# Create new regressor object
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor with highest p value
regressor_OLS.summary()


X_opt = X[:, [0, 3, 4, 5]]
# Create new regressor object
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor with highest p value
regressor_OLS.summary()


X_opt = X[:, [0, 3, 5]]
# Create new regressor object
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor with highest p value
regressor_OLS.summary()


X_opt = X[:, [0, 3]]
# Create new regressor object
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor with highest p value
regressor_OLS.summary()

File: 2-Regression/5-Multiple Linear Regression/multiple_linear_regression_grade_predict.R
# Multiple Linear Regression

# STEP 1

# Importing the dataset
gradesdataset = read.csv('50_Grades_Data.csv')

# Encoding categorical data
gradesdataset$Subject = factor(gradesdataset$Subject,
                               levels = c('English', 'Math'),
                               labels = c(1, 2))

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(gradesdataset$Subject, SplitRatio = 0.8)
training_set = subset(gradesdataset, split == TRUE)
test_set = subset(gradesdataset, split == FALSE)

# Manual feature scaling not required

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# STEP 2

# Fitting Multiple Linear Regression to the Training set

# regressor = lm(formula = Grade ~ MinutesStudy + HoursSleep + Subject)
# Short Version
regressor = lm(formula = Grade ~ .,
               data = training_set)

# Grade is linear combination of ind vars
# Type sumarry(regressor) in console
# Notice dummy vars were implemented automatically

# Lower p value means higher sinigicance of ind vars.

# In this example, the only strong predictor is Minutes Study
# So convert this to simple linear
# regressor = lm(formula = Grade ~ MinutesStudy, data = training_set)

# STEP 3

# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)

# Type y_pred in console to look at it, and compare tith test_set

# BACKWARD ELIMINATION

# Set significance level SL at 5%
# Building the optimal model using bacdward elmination
# Use regressor model but type out all ind vars
regressor = lm(formula = Grade ~ MinutesStudy + HoursSleep + Subject,
               data = gradesdataset) # on whole data set for complete info
# Beauty of R is no need to create dummy vars, the factor function does this when encoding categorical data

summary(regressor) # Returns data to search for vars with P value above signif level

# Complete backwared elimination, remove vars with high P values, first to remove is State
regressor = lm(formula = Grade ~ MinutesStudy + Subject,
               data = gradesdataset) # on whole data set for complete info
summary(regressor)

# Next, remove Subject
regressor = lm(formula = Grade ~ MinutesStudy,
               data = gradesdataset) # on whole data set for complete info
summary(regressor)